--- knot.orig/src/knot/common.h
+++ knot/src/knot/common.h
@@ -42,7 +42,7 @@ typedef unsigned int uint; /*!< \brief U
 
 #define PROJECT_EXEC SBINDIR "/" "knotd" /*!< \brief  Project executable. */
 #define ZONEPARSER_EXEC LIBEXECDIR "/" "knot-zcompile" /*!< \brief  Zoneparser executable. */
-#define PID_FILE "knot.pid" /*!< \brief Server PID file name. */
+#define PID_FILE "/run/knot/knotd.pid" /*!< \brief Server PID file name. */
 
 /*
  * Server.
--- knot.orig/src/knot/conf/conf.c
+++ knot/src/knot/conf/conf.c
@@ -154,7 +154,7 @@ static int conf_process(conf_t *conf)
 
 	// Create PID file
 	if (conf->pidfile == NULL) {
-		conf->pidfile = strcdup(conf->storage, "/" PID_FILE);
+		conf->pidfile = strdup(PID_FILE);
 		if (conf->pidfile == NULL) {
 			return KNOT_ENOMEM;
 		}
