From 2354047b6402aa68daffe96d6f82f30f0dad1cff Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Sur=C3=BD?= <ondrej@sury.org>
Date: Mon, 6 Jan 2014 11:37:23 +0100
Subject: [PATCH] Check the OpenSSL version when checking for GOST algorithm
 support

---
 src/libknot/dnssec/config.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- knot.orig/src/libknot/dnssec/config.h
+++ knot/src/libknot/dnssec/config.h
@@ -35,13 +35,13 @@
 #endif
 
 // ECDSA support requires OpenSSL version >= 1.0.1
-#if !defined(OPENSSL_NO_ECDSA) && OPENSSL_VERSION_NUMBER >= 0x10001000
+#if !defined(OPENSSL_NO_ECDSA) && OPENSSL_VERSION_NUMBER >= 0x10001000L
   #define KNOT_ENABLE_ECDSA 1
 #else
   #undef KNOT_ENABLE_ECDSA
 #endif
 
-#if !defined(OPENSSL_NO_GOST)
+#if !defined(OPENSSL_NO_GOST) && OPENSSL_VERSION_NUMBER >= 0x1000001fL
   #define KNOT_ENABLE_GOST 1
 #else
   #undef KNOT_ENABLE_GOST
