From: =?utf-8?q?Ond=C5=99ej_Sur=C3=BD?= <ondrej@debian.org>
Date: Tue, 16 Jun 2015 14:44:45 +0200
Subject: knot-doc-tweaks

---
 doc/Makefile.am | 28 +++++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/doc/Makefile.am b/doc/Makefile.am
index 652d200..e065fc6 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -15,7 +15,7 @@ EXTRA_DIST = \
 SPHINXBUILDDIR = _build
 ALLSPHINXOPTS  = -n -d $(SPHINXBUILDDIR)/doctrees -D latex_paper_size=a4 $(SPHINXOPTS) .
 
-.PHONY: html-local singlehtml pdf-local info-local
+.PHONY: html-local singlehtml pdf-local info-local install-html-local install-singlehtml install-pdf-local install-info-local
 
 if HAVE_SPHINXBUILD
 html-local:
@@ -23,19 +23,35 @@ html-local:
 	@echo
 	@echo "Build finished. The HTML pages are in $(SPHINXBUILDDIR)/html."
 
+install-html-local:
+	$(INSTALL) -d $(DESTDIR)/$(docdir) $(DESTDIR)/$(docdir)/_static $(DESTDIR)/$(docdir)/_sources
+	$(INSTALL) -D $(SPHINXBUILDDIR)/html/*.html $(DESTDIR)/$(docdir)/
+	$(INSTALL_DATA) $(SPHINXBUILDDIR)/html/_sources/* $(DESTDIR)/$(docdir)/_sources/
+	$(INSTALL_DATA) $(SPHINXBUILDDIR)/html/_static/* $(DESTDIR)/$(docdir)/
+
 singlehtml:
 	$(SPHINXBUILD) -b singlehtml $(ALLSPHINXOPTS) $(SPHINXBUILDDIR)/singlehtml
 	@echo
 	@echo "Build finished. The HTML page is in $(SPHINXBUILDDIR)/singlehtml."
 
+install-singlehtml:
+	$(INSTALL) -d $(DESTDIR)/$(docdir) $(DESTDIR)/$(docdir)/_static
+	$(INSTALL_DATA) $(SPHINXBUILDDIR)/singlehtml/*.html $(DESTDIR)/$(docdir)/
+	$(INSTALL_DATA) $(SPHINXBUILDDIR)/html/_static/* $(DESTDIR)/$(docdir)/_static/
+
 if HAVE_PDFLATEX
 pdf-local:
 	$(SPHINXBUILD) -b latex $(ALLSPHINXOPTS) $(SPHINXBUILDDIR)/latex
 	@echo "Running LaTeX files through pdflatex..."
 	$(MAKE) -C $(SPHINXBUILDDIR)/latex all-pdf
 	@echo "pdflatex finished; the PDF files are in $(SPHINXBUILDDIR)/latex."
+
+install-pdf-local:
+	$(INSTALL) -d $(DESTDIR)/$(docdir)
+	$(INSTALL_DATA) $(SPHINXBUILDDIR)/latex/* $(DESTDIR)/$(docdir)/
+
 else
-pdf:
+pdf-local install-pdf:
 	@echo "You need to install pdflatex and re-run configure to be"
 	@echo "able to generate PDF documentation."
 endif
@@ -46,13 +62,19 @@ info-local:
 	@echo "Running Texinfo files through makeinfo..."
 	make -C $(SPHINXBUILDDIR)/texinfo info
 	@echo "makeinfo finished; the Info files are in $(SPHINXBUILDDIR)/texinfo."
+
+install-info-local:
+	$(INSTALL) -d $(DESTDIR)/$(infodir)
+	$(INSTALL_DATA) $(SPHINXBUILDDIR)/texinfo/KnotDNS.info $(DESTDIR)/$(infodir)/knot.info
+
 else
+info-local install-info-local:
 	@echo "You need to install GNU Texinfo and re-run configure to be"
 	@echo "able to generate info pages."
 endif
 
 else
-html-local singlehtml pdf-local info-local:
+html-local singlehtml pdf-local info-local install-html-local install-singlehtml install-pdf-local install-info-local:
 	@echo "Please install sphinx (python-sphinx) to generate Knot DNS documentation."
 endif
 
