From: Ondřej Surý <ondrej@sury.org>
Date: Wed, 2 Sep 2015 10:22:05 +0200
Subject: knot-systemd-fixups

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 4197930..70faaf1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -101,7 +101,7 @@ AS_IF([test "$enable_systemd" != "no"],[
     [auto],[PKG_CHECK_MODULES([systemd], [libsystemd], [enable_systemd=yes], [
             PKG_CHECK_MODULES([systemd], [libsystemd-daemon libsystemd-journal], [enable_systemd=yes], [enable_systemd=no])])],
     [yes],[PKG_CHECK_MODULES([systemd], [libsystemd], [], [
-    →       PKG_CHECK_MODULES([systemd], [libsystemd-daemon libsystemd-journal])])],
+            PKG_CHECK_MODULES([systemd], [libsystemd-daemon libsystemd-journal])])],
     [*],[AC_MSG_ERROR([Invalid value of --enable-systemd.])])
     ])
 
