From 432027f7cd31be315d267e073e1034c9737fc419 Mon Sep 17 00:00:00 2001
From: Jan Vcelak <jan.vcelak@nic.cz>
Date: Mon, 15 Jul 2013 14:14:49 +0200
Subject: [PATCH] config: fix default configuration file load path

---
 src/knot/conf/conf.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- knot.orig/src/knot/conf/conf.c
+++ knot/src/knot/conf/conf.c
@@ -37,7 +37,7 @@
 
 /*! \brief Default config paths. */
 static const char *DEFAULT_CONFIG[] = {
-	SYSCONFDIR "/" "knot.conf",
+	CONFIG_DIR "/" "knot.conf",
 };
 
 #define DEFAULT_CONF_COUNT 1 /*!< \brief Number of default config paths. */
